src/Interactive/EditorFeatures/Core/Extensibility/Interactive/SendToInteractiveSubmissionProvider.cs

```diff
@@ -0,0 +1,103 @@
+
+namespace Microsoft.CodeAnalysis.Editor.Interactive
+{
+    /// <summary>
+    /// Implementers of this interface are responsible for retrieving source code that
+    /// should be sent to the REPL given the user's selection.
+    /// 
+    /// If the user does not make a selection then a line should be selected.
+    /// If the user selects code that fails to be parsed then the selection gets expanded
+    /// to a syntax node.
+    /// </summary>
+    internal abstract class SendToInteractiveSubmissionProvider : ISendToInteractiveSubmissionProvider
+    {
+        /// <summary>Expands the selection span of an invalid selection to a span that should be sent to REPL.</summary>
+        protected abstract IEnumerable<TextSpan> GetExecutableSyntaxTreeNodeSelection(TextSpan selectedSpan, SourceText source, SyntaxNode node, SemanticModel model);
+
+        /// <summary>Returns whether the submission can be parsed in interactive.</summary>
+        protected abstract bool CanParseSubmission(string code);
+
+        public string GetSelectedText(IEditorOptions editorOptions, CommandArgs args, CancellationToken cancellationToken)
+        {
+
+            IEnumerable<SnapshotSpan> selectedSpans = args.TextView.Selection.GetSnapshotSpansOnBuffer(args.SubjectBuffer).Where(ss => ss.Length > 0);
+
+            // If there is no selection select the current line.
+            if (!selectedSpans.Any())
+            {
+                selectedSpans = GetSelectedLine(args);
+            }
+
+            // Send the selection as is if it does not contain any parsing errors.
+            var candidateSubmission = GetSubmissionFromSelectedSpans(editorOptions, selectedSpans);
+            if (CanParseSubmission(candidateSubmission))
+            {
+                return candidateSubmission;
+            }
+
+            // Otherwise heuristically try to expand it.
+            return GetSubmissionFromSelectedSpans(editorOptions, ExpandSelection(selectedSpans, args, cancellationToken));
+        }
+
+        /// <summary>Returns the span for the currently selected line.</summary>
+        private static IEnumerable<SnapshotSpan> GetSelectedLine(CommandArgs args)
+        {
+            SnapshotPoint? caret = args.TextView.GetCaretPoint(args.SubjectBuffer);
+            int caretPosition = args.TextView.Caret.Position.BufferPosition.Position;
+            ITextSnapshotLine containingLine = caret.Value.GetContainingLine();
+            return new SnapshotSpan[] {
+                new SnapshotSpan(containingLine.Start, containingLine.End)
+            };
+        }
+
+        private async Task<IEnumerable<SnapshotSpan>> GetExecutableSyntaxTreeNodeSelection(
+            TextSpan selectionSpan,
+            CommandArgs args,
+            ITextSnapshot snapshot,
+            CancellationToken cancellationToken)
+        {
+            Document doc = args.SubjectBuffer.GetRelatedDocuments().FirstOrDefault();
```

Comment:
Oh, I see. Yes, in case of defines and use of linked files the selection might not have had been correctly expanded.
I updated code to pull currently relevant Document from the workspace.
Thanks.